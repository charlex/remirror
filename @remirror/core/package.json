{
  "name": "@remirror/core",
  "description": "Provides the core abstractions for managing your remirror editor functionality",
  "license": "MIT",
  "author": "Ifiok Jr. <ifiokotung@gmail.com>",
  "homepage": "https://github.com/ifiokjr/remirror/tree/master/@remirror/core",
  "repository": "https://github.com/ifiokjr/remirror/tree/master/@remirror/core",
  "version": "0.4.2-canary.2",
  "main": "lib/index.js",
  "module": "lib/dist/core.esm.js",
  "browser": {
    "min-document": false
  },
  "files": [
    "internal",
    "lib",
    "src"
  ],
  "scripts": {
    "api": "api-extractor run",
    "api:local": "api-extractor run --local",
    "build": "run-p build:*",
    "build:babel": "babel src --out-dir lib --root-mode=upward --extensions \".ts,.tsx\" -s",
    "build:declaration": "tsc -p ./tsconfig.prod.json --emitDeclarationOnly --declarationMap",
    "dev": "run-p dev:*",
    "dev:babel": "babel src --out-dir lib --root-mode=upward --extensions \".ts,.tsx\" -s --watch",
    "dev:declaration": "tsc -p ./tsconfig.prod.json --emitDeclarationOnly --declarationMap --watch",
    "document": "node ../api-documenter/lib/start.js markdown --input-folder temp --output-folder ../../docs/api/@remirror",
    "lint": "tslint --project tsconfig.lint.json --config ../../tslint.json",
    "typecheck": "tsc -p ./tsconfig.json --noEmit"
  },
  "types": "lib/index.d.ts",
  "dependencies": {
    "@babel/runtime": "^7.6.0",
    "@remirror/core-constants": "0.4.2-canary.2",
    "@remirror/core-helpers": "0.4.2-canary.2",
    "@remirror/core-types": "0.4.2-canary.2",
    "@remirror/core-utils": "0.4.2-canary.2",
    "@remirror/react-portals": "0.4.2-canary.2",
    "@types/prosemirror-inputrules": "^1.0.2",
    "@types/prosemirror-keymap": "^1.0.1",
    "@types/prosemirror-model": "^1.7.2",
    "@types/prosemirror-state": "^1.2.3",
    "prosemirror-inputrules": "^1.0.4",
    "prosemirror-keymap": "^1.0.1",
    "prosemirror-model": "^1.7.2",
    "prosemirror-state": "^1.2.4",
    "prosemirror-suggest": "0.4.2-canary.2"
  },
  "peerDependencies": {
    "prosemirror-view": "^1.11.1",
    "@types/prosemirror-view": "^1.9.0",
    "@emotion/core": "^10",
    "@types/react": "^16.8.0",
    "react": "^16.8.0"
  },
  "publishConfig": {
    "access": "public"
  },
  "cjs": "lib/dist/core.cjs.js",
  "meta": {
    "sizeLimit": "55 KB"
  },
  "sideEffects": false
}
