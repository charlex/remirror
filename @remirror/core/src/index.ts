export * from './extension';
export * from './extension-helpers';
export * from './extension-manager';
export * from './extension-types';
export * from './mark-extension';
export * from './node-extension';
export * from './nodes';

export * from '@remirror/core-constants';
export * from '@remirror/core-types';
export * from '@remirror/core-utils';
export * from '@remirror/core-helpers';
