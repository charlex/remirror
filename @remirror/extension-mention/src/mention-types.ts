import {
  Attrs,
  EditorViewParams,
  FromToParams,
  MakeRequired,
  MarkExtensionOptions,
  MarkTypeParams,
} from '@remirror/core';

export interface OptionalMentionExtensionParams {
  /**
   * The text to append to the replacement.
   *
   * @defaultValue '''
   */
  appendText?: string;

  /**
   * The type of replacement to use. By default the command will only replace text up the the cursor position.
   *
   * To force replacement of the whole match regardless of where in the match the cursor is placed set this to
   * `full`.
   *
   * @defaultValue 'full'
   */
  replacementType?: SuggestionReplacementType;
}

/**
 * The attrs that will be added to the node.
 * ID and label are plucked and used while attributes like href and role can be assigned as desired.
 */
export type MentionExtensionAttrs = Attrs<
  OptionalMentionExtensionParams & {
    /**
     * A unique identifier for the suggestions node
     */
    id: string;

    /**
     * The text to be placed within the suggestions node
     */
    label: string;

    /**
     * The name of the matched char
     */
    name?: string;

    /**
     * The range of the requested selection.
     */
    range?: FromToEndParams;
  }
>;

export interface SuggestionMatcher {
  /**
   * The character to match against
   *
   * @defaultValue '@'
   */
  char: string;

  /**
   * Whether to only match from the start of the line
   *
   * @defaultValue `false`
   */
  startOfLine: boolean;

  /**
   * A regex containing all supported characters.
   *
   * @defaultValue `/[\w\d_]+/`
   */
  supportedCharacters: RegExp | string;

  /**
   * Name of matching character - This will be appended to the classnames
   *
   * @defaultValue 'at'
   */
  name: string;
}

export type OptionalSuggestionMatcher = MakeRequired<Partial<SuggestionMatcher>, 'char' | 'name'>;

export interface MatchLength {
  /**
   * The complete match independent of the cursor position.
   */
  full: string;

  /**
   * This value is a partial match which ends at the position of the cursor within the matching text.
   */
  partial: string;
}

/**
 *
 */
export interface FromToEndParams extends FromToParams {
  /**
   * The absolute end of the matching string.
   */
  end: number;
}

export interface SuggestionStateMatch
  extends Pick<SuggestionMatcher, 'name' | 'char'>,
    SuggestionMatcherParams {
  /**
   * Range of current match.
   * - `from` is the start
   * - `to` is cursor position
   * - `end` is the end of the match
   */
  range: FromToEndParams;

  /**
   * Current query of match which doesn't include the match character.
   */
  query: MatchLength;

  /**
   * Full text of match including the activation character e.g. `'@awesome'`.
   */
  text: MatchLength;

  /**
   * The characters that the matcher supports
   */
  supportedCharacters: RegExp;
}

export interface SuggestionMatchParams {
  /**
   * The match that will be triggered.
   */
  match: SuggestionStateMatch;
}

/**
 * Determines whether to replace the full match or the partial match (up to the cursor position).
 */
export type SuggestionReplacementType = 'full' | 'partial';

export interface SuggestionCommandParams extends MarkTypeParams, OptionalMentionExtensionParams {
  attrs: MentionExtensionAttrs;
  range: FromToEndParams;
}

export interface SuggestionCallbackParams extends SuggestionStateMatch, EditorViewParams, SuggestionActions {}

export interface OnKeyDownParams extends SuggestionStateMatch, EditorViewParams {
  /**
   * The keyboard event
   */
  event: KeyboardEvent;
}

export interface ReasonParams<GReason> {
  /**
   * The reason this callback was triggered. This can be used to determine the action to use in your own code.
   */
  reason: GReason;
}

export interface OnChangeCallbackParams extends SuggestionCallbackParams, ReasonParams<ChangeReason> {}

export interface OnExitCallbackParams extends SuggestionCallbackParams, ReasonParams<ExitReason> {}

/**
 * The options passed into a mention
 */
export interface MentionExtensionOptions extends MarkExtensionOptions {
  /**
   * Provide customs class names for the completed mention
   */
  mentionClassName?: string;

  /**
   * Provide a custom tag for the mention
   */
  readonly tag?: keyof HTMLElementTagNameMap;

  /**
   * Provide a custom matcher with options
   */
  matchers?: OptionalSuggestionMatcher[];

  /**
   * Text to append after the mention has been added.
   *
   * @defaultValue ' '
   *
   * TODO move this into the `matchers` option as a property of the SuggestionMatcher
   */
  appendText?: string;

  /**
   * Class name to use for the decoration (while the plugin is still being written)
   */
  suggestionClassName?: string;

  /**
   * Tag which wraps an active match
   */
  decorationsTag?: keyof HTMLElementTagNameMap;

  /**
   * Called when a suggestion is active and has changed.
   *
   * @defaultValue `() => void`
   */
  onChange?(params: OnChangeCallbackParams): void;

  /**
   * Called when a suggestion is exited with the pre-exit match.
   * Can be used to force the command to run e.g. when no match was found but a hash
   * should still be created this can be used to call the command parameter and trigger the mention being created.
   *
   * @defaultValue `() => void`
   */
  onExit?(params: OnExitCallbackParams): void;

  /**
   * Called for each character entry and can be used to disable certain characters.
   *
   * @remarks
   * For example you may want to disable all `@` symbols while the `at` matcher is active.
   * Return true to stop the default character action.
   *
   * @defaultValue `() => false`
   */
  onCharacterEntry?(params: OnCharacterEntryParams): boolean;

  /**
   * An object that describes how certain key bindings should be handled.
   *
   * Return `true` to prevent any further prosemirror actions or return `false` to allow prosemirror to continue.
   */
  keyBindings?: SuggestionKeyBindingMap;
}

export interface OnCharacterEntryParams extends SuggestionCallbackParams {
  /**
   * The text entry that was received.
   */
  entry: FromToParams & { text: string };
}

export interface SuggestionKeyBindingParams extends SuggestionCallbackParams {
  /**
   * The native keyboard event.
   */
  event: KeyboardEvent;
}

/**
 * A method for performing actions when a certain key / patter is pressed.
 *
 * Return true to prevent any further bubbling of the key event and to stop other handlers
 * from also processing the event.
 */
export type SuggestionKeyBinding = (params: SuggestionKeyBindingParams) => boolean | void;

/**
 * The handler callback signature.
 */
export type SuggestionCallback = (params: SuggestionCallbackParams) => void;

/**
 * The SuggestionKeyBinding object.
 */
export type SuggestionKeyBindingMap = Partial<
  Record<'Enter' | 'ArrowDown' | 'ArrowUp' | 'Esc' | 'Delete' | 'Backspace', SuggestionKeyBinding>
> &
  Record<string, SuggestionKeyBinding>;

/**
 * A suggestion has two stage. When it is `new` and when it has already been created
 * and is now being `edit`ed.
 *
 * Separating the stages allows for greater control in how mentions are updated.
 */
export type SuggestionStage = 'new' | 'edit';

export type SuggestionCommandAttrs = Attrs<
  Partial<Pick<MentionExtensionAttrs, 'id' | 'label' | 'appendText' | 'replacementType'>>
>;
export interface SuggestionActions {
  /**
   * Create a mention for the first time
   */
  create(attrs: MentionExtensionAttrs): void;

  /**
   * Update the active mention
   */
  update(attrs: MentionExtensionAttrs): void;

  /**
   * Remove the active suggestion
   */
  remove(attrs: Attrs<FromToParams>): void;

  /**
   * A command automatically provides some default.
   *
   * This is the recommended option to use.
   */
  command(attrs: SuggestionCommandAttrs): void;

  /**
   * The current stage of the focused mention. Can be used to decide which action to use.
   */
  stage: SuggestionStage;
}

/**
 * The action taken on a suggestion
 */
export enum ActionTaken {
  /**
   * We've moved from suggestion to another.
   */
  Moved = 'moved',

  /**
   * The suggestion wasn't active before and now it is active
   */
  Entered = 'entered',

  /**
   * The suggestion was active before and not it is no longer active
   */
  Exited = 'exited',

  /**
   * The suggestion query has changed either by typing or deleting characters
   */
  Changed = 'changed',
}

/**
 * The potential reasons for an exit
 */
export enum ExitReason {
  /**
   * The user has pasted some text with multiple characters or run a command that adds multiple character.
   *
   * `onExit` should be called but the previous match should be retested as it's possible that it's been extended.
   */
  End = 'exit-end',

  /**
   * The suggestion has been removed.
   */
  Removed = 'delete',

  /**
   * The user has pasted some text with multiple characters or run a command that adds multiple characters
   * somewhere within the active suggestion.
   * e.g. `@abc` -> `@ab123 asdf aiti c`
   *
   * `onExit` should be called but the previous match should be retested as it's possible that it's been extended.
   */
  Split = 'exit-split',

  /**
   * The user has pasted some text with multiple characters or run a command that adds multiple characters
   * right after the initial multi-character.
   * e.g. `@abc` -> `@ this is newabc`
   *
   * In this case it is best to remove the mention completely.
   */
  InvalidSplit = 'invalid-exit-split',

  /**
   * User has moved out of the suggestion at the end. This will typically be using arrow keys, but can also be
   * via a mouse click or custom command. All that has changed is the cursor position.
   */
  MoveEnd = 'move-end',

  /**
   * User has moved out of the suggestion but from the beginning. This can be via the arrow keys but can also be
   * via a mouse click or custom command. All that changed is the cursor position.
   */
  MoveStart = 'move-start',

  /**
   * The user has jumped to another suggestion which occurs afterwards in the editor. This can be via a click,
   * a keyboard jump or custom commands. In this case since there is still an active suggestion it will trigger
   * both an `onExit` and `onChange` call.
   */
  JumpForward = 'jump-forward-exit',

  /**
   * The user has jumped to another suggestion which occurs before the previous suggestion in the editor. This can
   * happen via a click, a keyboard jump (END) or a custom command. In this case since there is still an active
   * suggestion it will trigger both an `onExit` and `onChange` call.
   */
  JumpBackward = 'jump-backward-exit',

  /**
   * The user has selected some text outside the current selection, this can trigger
   * an exit. This can be from a triple click to select the line or Ctrl-A to select all.
   */
  SelectionOutside = 'selection-outside',
}

/**
 * The potential reason for changes
 */
export enum ChangeReason {
  /**
   * The user has entered or started a new suggestion.
   */
  Start = 'start',

  /**
   * A changed happened to the character. This can be addition, deletion or replacement.
   */
  Text = 'change-character',

  /**
   * A change happened to the selection status.
   */
  SelectionInside = 'selection-inside',

  /**
   * The cursor was moved.
   */
  Move = 'move',

  /**
   * The user has moved from one suggestion to another suggestion earlier in the document.
   */
  JumpBackward = 'jump-backward-change',
  /**
   * The user has moved from one suggestion to another suggestion further along in the document.
   */
  JumpForward = 'jump-forward-change',
}

export interface SuggestionMatcherParams {
  /**
   * The matcher to search through.
   */
  matcher: Partial<SuggestionMatcher>;
}

export interface SuggestionStateMatchReason<GReason> extends SuggestionStateMatch, ReasonParams<GReason> {}

export interface SuggestionReasonMap {
  /**
   * Reasons triggering the onChange handler.
   */
  change?: SuggestionStateMatchReason<ChangeReason>;

  /**
   * Reasons triggering the `onExit` handler
   */
  exit?: SuggestionStateMatchReason<ExitReason>;
}

/**
 * Compares two matches.
 */
export interface CompareMatchParams {
  /**
   * The initial match
   */
  prev: SuggestionStateMatch;

  /**
   * The current match
   */
  next: SuggestionStateMatch;
}

export interface ReasonMatchParams<GReason> {
  /**
   * The match with its reason property.
   */
  match: SuggestionStateMatchReason<GReason>;
}

/**
 * Compares two matches.
 */
export interface CompareMatchParams {
  /**
   * The initial match
   */
  prev: SuggestionStateMatch;

  /**
   * The current match
   */
  next: SuggestionStateMatch;
}

export type MentionExtensionCommands = 'createMention' | 'updateMention' | 'removeMention';
