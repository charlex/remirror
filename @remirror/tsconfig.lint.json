{
  "extends": "../support/tsconfig.base.json",
  "compilerOptions": {
    "declaration": false,
    "allowJs": true,
    "skipLibCheck": true,
    "types": [
      "jest",
      "jest-extended",
      "jest-axe",
      "@testing-library/jest-dom/extend-expect",
      "jest-emotion",
      "jest-enzyme",
      "jest-environment-enzyme"
    ]
  },
  "include": ["**/*.ts", "**/*.tsx", "**/*.js"],
  "exclude": ["lib", "**/lib", "node_modules/**"]
}
