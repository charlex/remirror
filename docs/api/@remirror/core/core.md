<!-- Do not edit this file. It is automatically generated by API Documenter. -->

[Home](./index.md) &gt; [@remirror/core](./core.md)

## core package

## Classes

|  Class | Description |
|  --- | --- |
|  [DocExtension](./core.docextension.md) | The parent node in a Prosemirror Schema. A representation of the doc is required in all editors  |
|  [Extension](./core.extension.md) | Extensions are fundamental to the way that Remirror works and they handle the management of similar concerns. |
|  [ExtensionManager](./core.extensionmanager.md) | A class to manage the extensions and prosemirror interactions of our editor. |
|  [MarkExtension](./core.markextension.md) | All mark extensions should be created from here. |
|  [NodeExtension](./core.nodeextension.md) | Defines an abstract class for creating extensions which can place nodes into the prosemirror state. |
|  [TextExtension](./core.textextension.md) | The default text passed into the prosemirror schema. |

## Interfaces

|  Interface | Description |
|  --- | --- |
|  [DocExtensionOptions](./core.docextensionoptions.md) |  |
|  [Extension](./core.extension.md) |  |
|  [ExtensionConstructor](./core.extensionconstructor.md) | An extension constructor |
|  [ExtensionListParams](./core.extensionlistparams.md) |  |
|  [ExtensionManagerData](./core.extensionmanagerdata.md) |  |
|  [ExtensionParams](./core.extensionparams.md) |  |
|  [ManagerParams](./core.managerparams.md) |  |
|  [OnTransactionManagerParams](./core.ontransactionmanagerparams.md) |  |
|  [PrioritizedExtension](./core.prioritizedextension.md) | Provides a priority value to the extension which determines the priority. |

## Variables

|  Variable | Description |
|  --- | --- |
|  [isExtension](./core.isextension.md) | Determines if the passed in extension is a any type of extension. |
|  [isExtensionManager](./core.isextensionmanager.md) | Checks to see whether this is an extension manager |
|  [isMarkExtension](./core.ismarkextension.md) | Determines if the passed in extension is a mark extension. Useful as a type guard where a particular type of extension is needed. |
|  [isNodeExtension](./core.isnodeextension.md) | Determines if the passed in extension is a node extension. Useful as a type guard where a particular type of extension is needed. |
|  [isPlainExtension](./core.isplainextension.md) | Checks whether the this is an extension and if it is a plain one |

## Type Aliases

|  Type Alias | Description |
|  --- | --- |
|  [ActionNames](./core.actionnames.md) | Utility type for pulling all the action names from a list |
|  [ActionsFromExtensions](./core.actionsfromextensions.md) | Utility type which receives an extension and provides the type of actions it makes available. |
|  [AnyExtension](./core.anyextension.md) | Provides a type annotation which is applicable to any extension type. |
|  [AnyExtensionManager](./core.anyextensionmanager.md) | The utility for capturing all the possible <code>ExtensionManager</code> type variations. |
|  [CommandsOfExtension](./core.commandsofextension.md) | Utility type for retrieving the commands provided by an extension. |
|  [DataOfExtension](./core.dataofextension.md) | Utility type for retrieving the extension data object made available from the extension. |
|  [ExtensionsFromManager](./core.extensionsfrommanager.md) | Retrieve the extensions from an <code>ExtensionManager</code>. |
|  [FlexibleExtension](./core.flexibleextension.md) | Either a PrioritizedExtension or the actual Extension. |
|  [FlexibleExtensions](./core.flexibleextensions.md) |  |
|  [FromFlexibleExtension](./core.fromflexibleextension.md) |  |
|  [HelperNames](./core.helpernames.md) | Utility type for pulling all the action names from a list |
|  [HelpersFromExtensions](./core.helpersfromextensions.md) | Utility type which receives an extension and provides the type of helpers it makes available. |
|  [HelpersOfExtension](./core.helpersofextension.md) | Utility type for retrieving the helpers provided by an extension. |
|  [InferFlexibleExtension](./core.inferflexibleextension.md) | Get the extension from a PrioritizedExtension. |
|  [InferFlexibleExtensionList](./core.inferflexibleextensionlist.md) | Get the extension types from a list of PrioritizedExtensions. |
|  [MapCommandToAction](./core.mapcommandtoaction.md) | A utility type which maps the passed in extension command in an action that is called via <code>manager.data.actions.commandName()</code>. |
|  [MapHelpers](./core.maphelpers.md) | A utility type which maps the passed in extension helpers to a method called with <code>manager.data.helpers.helperName()</code>. |
|  [MarkNames](./core.marknames.md) | A utility type for retrieving the name of an extension only when it's a mark extension. |
|  [NameOfExtension](./core.nameofextension.md) | Utility type for retrieving the name of an extension. |
|  [NodeNames](./core.nodenames.md) | A utility type for retrieving the name of an extension only when it's a node extension. |
|  [OptionsOfExtension](./core.optionsofextension.md) | Utility type for retrieving the extension options from an extension. |
|  [PlainNames](./core.plainnames.md) | A utility type for retrieving the name of an extension only when it's a plain extension. |
|  [ProsemirrorTypeOfExtension](./core.prosemirrortypeofextension.md) | Utility type for retrieving the prosemirror type of the extension which can be:<!-- -->- <code>NodeType</code> - <code>MarkType</code> - <code>never</code> |
|  [SchemaFromExtensions](./core.schemafromextensions.md) | Gets the schema from a list of extensions |
|  [TypeOfExtensionClass](./core.typeofextensionclass.md) | Retrieve the instance type from an ExtensionClass. |
|  [TypeOfExtensionClassList](./core.typeofextensionclasslist.md) | This is a utility type which allows for retrieving the instance types from an array of Extensions. |

