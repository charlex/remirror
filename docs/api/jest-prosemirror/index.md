<!-- Do not edit this file. It is automatically generated by API Documenter. -->

[Home](./index.md)

## API Reference

## Packages

|  Package | Description |
|  --- | --- |
|  [jest-prosemirror](./jest-prosemirror.md) | \#\# The problem<!-- -->You want to write tests for some of your prosemirror editor but you don't know where to start. You know you should avoid testing implementation details and just want to be sure that your commands and plugins produce the correct underlying prosemirror state.<!-- -->\#\# This solution<code>jest-prosemirror</code> takes inspiration from the \[<code>testing-library</code>\](https://github.com/testing-library/react-testing-library) mantra and enables you to write more intuitive tests for your prosemirror editor.<!-- -->\#\# Installation
```bash
yarn add jest-prosemirror

```
\#\# Getting started<!-- -->\#\#\# Quick setup<!-- -->For a quick setup add the following to your jest.config.js file.
```js
module.exports = {
setupFilesAfterEnv: ['jest-prosemirror/environment'],
testEnvironment: 'jsdom', // Required for dom manipulation
};

```
This will automatically:<!-- -->- Add the jest assertions <code>toTransformNode</code> and <code>toEqualProsemirrorNode</code>.<!-- -->If you are using typescript then add this to your <code>tsconfig.json</code> file for global type support.
```json
{
"compilerOptions": {
 "types": ["jest-prosemirror"]
}
}

```
\#\#\# Manual setup<!-- -->Create a <code>jest.framework.dom.ts</code> file and add the following
```ts
import { prosemirrorMatchers } from 'jest-prosemirror';

// Add jest-prosemirror assertions
expect.extend(prosemirrorMatchers);

```
In your <code>jest.config.js</code> add the created file to your configuration.
```js
module.exports = {
setupFilesAfterEnv: ['<rootDir>/jest.framework.dom.ts'],
testEnvironment: 'jsdom', // Required for dom manipulation
};

```
\#\# Snapshot serializer<!-- -->This package exports a serializer for better snapshot testing of prosemirror primitives. To set this up add the following to your <code>jest.config.js</code> file.
```js
module.exports = {
snapshotSerializers: ['jest-prosemirror/serializer'],
};

```
Alternatively, you can add the following to your <code>jest.framework.dom.ts</code> file.
```ts
import { prosemirrorSerializer } from 'jest-prosemirror';

// Add the serializer for use throughout all the configured test files.
expect.addSnapshotSerializer(prosemirrorSerializer);

```
 |

