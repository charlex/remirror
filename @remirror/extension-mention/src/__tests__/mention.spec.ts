import { fromHTML, toHTML } from '@remirror/core';
import { createBaseTestManager } from '@test-fixtures/schema-helpers';
import { pmBuild } from 'jest-prosemirror';
import { renderEditor } from 'jest-remirror';
import { MentionExtension, MentionExtensionOptions } from '../';
import { SuggestionCallbackParams } from '../mention-types';

describe('schema', () => {
  const { schema } = createBaseTestManager([{ extension: new MentionExtension(), priority: 1 }]);
  const attrs = { id: 'test', label: '@test' };

  const { mention, p, doc } = pmBuild(schema, {
    mention: { markType: 'mention', ...attrs },
  });

  it('creates the correct dom node', () => {
    expect(toHTML({ node: p(mention(attrs.label)), schema })).toBe(
      `<p><a class="mention mention-at" data-mention-id="${attrs.id}">${attrs.label}</a></p>`,
    );
  });

  it('parses the dom structure and finds itself', () => {
    const node = fromHTML({
      schema,
      content: `<a class="mention mention-at" data-mention-id="${attrs.id}">${attrs.label}</a>`,
    });
    const expected = doc(p(mention(attrs.label)));
    expect(node).toEqualProsemirrorNode(expected);
  });
});

describe('constructor', () => {
  it('is created with the correct options', () => {
    const matcher = {
      char: '@',
      allowSpaces: false,
      startOfLine: false,
      name: 'at',
    };
    const mentions = new MentionExtension({
      matchers: [matcher],
    });

    expect(mentions.options.matchers).toEqual([matcher]);
    expect(mentions.name).toEqual('mention');
  });

  it('uses can be created with partial matchers', () => {
    const mentionOne = new MentionExtension({
      matchers: [{ char: '#', name: 'tag' }],
    });
    expect(mentionOne.options.matchers).toEqual([{ char: '#', name: 'tag' }]);
  });
});

const create = (params: MentionExtensionOptions = {}) =>
  renderEditor({
    attrMarks: [new MentionExtension({ mentionClassName: 'custom', ...params })],
  });

describe('plugin', () => {
  const options = {
    mentionClassName: 'custom',
    matchers: [{ char: '#', name: 'tag' }, { char: '@', name: 'at' }, { char: '+', name: 'plus' }],
  };

  const mocks = {
    onChange: jest.fn(),
    onExit: jest.fn(({ command }: SuggestionCallbackParams) => {
      command({ appendText: '' });
    }),
  };

  const id = 'mention';
  const label = `@${id}`;

  let {
    add,
    nodes: { doc, p },
    attrMarks: { mention },
    view,
  } = create({
    ...options,
    ...mocks,
  });
  let mentionMark = mention({ id, label });

  beforeEach(() => {
    ({
      add,
      nodes: { doc, p },
      attrMarks: { mention },
      view,
    } = create({
      ...options,
      ...mocks,
    }));
    mentionMark = mention({ id, label });
  });

  it('uses default noop callbacks', () => {
    ({
      add,
      nodes: { doc, p },
      attrMarks: { mention },
      view,
    } = create(options));

    add(doc(p('<cursor>'))).insertText(`This ${label} `);
    expect(view.state).toContainRemirrorDocument(p(`This ${label} `));
  });

  it('should support onExit', () => {
    add(doc(p('<cursor>'))).insertText(`${label} `);
    expect(view.state).toContainRemirrorDocument(p(mentionMark(label), ' '));
  });

  it('should handle joined text separated by space', () => {
    add(doc(p('hello <cursor>friend'))).insertText(`${label} `);
    expect(view.state).toContainRemirrorDocument(p('hello ', mentionMark(label), ' friend'));
  });

  it('can split mentions', () => {
    const splitMention = mention({ id: '123', label: '@123' });
    add(doc(p(splitMention('@1<cursor>23')))).insertText(` `);
    expect(view.state).toContainRemirrorDocument(p(mention({ id: '1', label: '@1' })('@1'), ' 23'));
  });

  it('removes invalid mentions', () => {
    const splitMention = mention({ id: '123', label: '@123' });
    add(doc(p(splitMention('@<cursor>123')))).insertText(` `);
    expect(view.state).toContainRemirrorDocument(p('@ 123'));
  });

  it('decorates split mentions', () => {
    add(doc(p('hello <cursor>friend'))).insertText(`${label}`);
    expect(view.dom).toContainHTML('<a class="suggestion suggestion-at">@mentionfriend</a>');
  });

  it('injects the mention at the correct place', () => {
    add(doc(p('<cursor>'))).insertText(`This ${label} `);
    expect(view.state).toContainRemirrorDocument(p('This ', mentionMark(label), ' '));
    expect(mocks.onChange).toHaveBeenCalledTimes(id.length);
  });

  it('supports multiple characters', () => {
    const labelFn = (char: string) => `${char}${id}`;

    const hashMark = mention({ id, label: labelFn('#'), name: 'tag' });
    const plusMark = mention({ id, label: labelFn('+'), name: 'plus' });
    const atMark = mention({ id, label: labelFn('@'), name: 'at' });

    add(doc(p('<cursor>'))).insertText(`This ${labelFn('#')} `);
    expect(view.state).toContainRemirrorDocument(p('This ', hashMark(labelFn('#')), ' '));

    add(doc(p('<cursor>'))).insertText(`This ${labelFn('+')} `);
    expect(view.state).toContainRemirrorDocument(p('This ', plusMark(labelFn('+')), ' '));

    add(doc(p('<cursor>'))).insertText(`This ${labelFn('@')} `);
    expect(view.state).toContainRemirrorDocument(p('This ', atMark(labelFn('@')), ' '));
  });
});

describe('commands', () => {
  const options = {
    matchers: [{ char: '#', name: 'tag' }, { char: '@', name: 'at' }, { char: '+', name: 'plus' }],
  };
  let {
    nodes: { doc, p },
    view,
    attrMarks: { mention },
    actions,
    add,
  } = create(options);

  const attrs = { id: 'test', label: '@test', name: 'at', appendText: '' };

  beforeEach(() => {
    ({
      nodes: { doc, p },
      view,
      attrMarks: { mention },
      actions,
      add,
    } = create(options));
  });

  describe('createMention', () => {
    it('replaces text at the current position by default', () => {
      add(doc(p('This is ', '<cursor>')));
      actions.createMention(attrs);

      expect(view.state).toContainRemirrorDocument(p('This is ', mention(attrs)(attrs.label)));
    });

    it('replaces text at the specified position', () => {
      add(doc(p('This is ', '<cursor>')));
      actions.createMention({ ...attrs, range: { from: 1, to: 1, end: 1 } });

      expect(view.state).toContainRemirrorDocument(p(mention(attrs)(attrs.label), 'This is '));
    });

    it('throws when invalid config passed into the command', () => {
      add(doc(p('This is ', '<cursor>')));

      expect(() => actions.createMention()).toThrowErrorMatchingInlineSnapshot(
        `"Invalid configuration attributes passed to the MentionExtension command."`,
      );
      expect(() => actions.createMention({})).toThrowErrorMatchingInlineSnapshot(
        `"Invalid configuration attributes passed to the MentionExtension command."`,
      );
      expect(() => actions.createMention({ ...attrs, id: '' })).toThrowErrorMatchingInlineSnapshot(
        `"Invalid configuration attributes passed to the MentionExtension command."`,
      );
      expect(() => actions.createMention({ ...attrs, label: '' })).toThrowErrorMatchingInlineSnapshot(
        `"Invalid configuration attributes passed to the MentionExtension command."`,
      );
      expect(() => actions.createMention({ ...attrs, name: 'invalid' })).toThrowErrorMatchingInlineSnapshot(
        `"The name 'invalid' specified for this command is invalid. Please choose from: [\\"tag\\",\\"at\\",\\"plus\\"]."`,
      );
      expect(() => actions.createMention({ ...attrs, name: undefined })).toThrowErrorMatchingInlineSnapshot(
        `"The MentionExtension command must specify a name since there are multiple matchers configured"`,
      );
    });
  });
});
