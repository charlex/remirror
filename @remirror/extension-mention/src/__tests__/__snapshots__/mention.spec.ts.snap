// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin supports deleting content 1`] = `
Prosemirror node: {
  "type": "doc",
  "content": [
    {
      "type": "paragraph",
      "attrs": {
        "align": null,
        "id": null,
        "indent": 0,
        "lineSpacing": null
      }
    }
  ]
}
`;
